controllers/appmesh/cloudmap_controller.go

```diff
@@ -0,0 +1,888 @@
+	virtualNodeName string
+	meshName        string
+	healthStatus    string
+	labels          map[string]string
+}
+
+type cloudmapNamespaceCacheItem struct {
+	key   string
+	value cloudMapNamespaceSummary
+}
+
+type cloudMapNamespaceSummary struct {
+	NamespaceID   string
+	NamespaceType string
+}
+
+type cloudmapServiceCacheItem struct {
+	key   string
+	value cloudMapServiceSummary
+}
+
+type cloudMapServiceSummary struct {
+	NamespaceID string
+	ServiceID   string
+}
+
+var cloudMapBackoff = wait.Backoff{
+	Steps:    4,
+	Duration: 15 * time.Second,
+	Factor:   1.0,
+	Jitter:   0.1,
+	Cap:      60 * time.Second,
+}
+
+func NewCloudMapReconciler(k8sClient client.Client, finalizerManager k8s.FinalizerManager, cloudMapSDK awsservices.CloudMap, log logr.Logger) *cloudMapReconciler {
+	return &cloudMapReconciler{
+		k8sClient:        k8sClient,
+		log:              log,
+		cloudMapSDK:      cloudMapSDK,
+		finalizerManager: finalizerManager,
+		cloudMapInstanceCache: cache.NewTTLStore(func(obj interface{}) (string, error) {
+			return obj.(*cloudMapInstanceCacheItem).key, nil
+		}, 300*time.Second),
+		nameSpaceIDCache: cache.NewTTLStore(func(obj interface{}) (string, error) {
+			return obj.(*cloudmapNamespaceCacheItem).key, nil
+		}, 60*time.Second),
+		serviceIDCache: cache.NewTTLStore(func(obj interface{}) (string, error) {
+			return obj.(*cloudmapServiceCacheItem).key, nil
+		}, 60*time.Second),
+		enqueueRequestsForPodEvents: virtualnode.NewEnqueueRequestsForPodEvents(k8sClient, log),
+	}
+}
+
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes,verbs=get;list;watch
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+
+func (r *cloudMapReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	return runtime.HandleReconcileError(r.reconcile(req), r.log)
+}
+
+func (r *cloudMapReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appmesh.VirtualNode{}).
+		Watches(&source.Kind{Type: &corev1.Pod{}}, r.enqueueRequestsForPodEvents).
+		Complete(r)
+}
+
+func (r *cloudMapReconciler) reconcile(req ctrl.Request) error {
+	ctx := context.Background()
+	log := r.log.WithValues("Virtualnode", req.NamespacedName)
+
+	vNode := &appmesh.VirtualNode{}
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, vNode); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	log.V(2).Info("VNode: ", "awsName: ", vNode.Spec.AWSName,
+		"vNode.PodSelector: ", vNode.Spec.PodSelector)
+	if !vNode.DeletionTimestamp.IsZero() {
+		return r.deleteCloudMapResources(ctx, vNode)
+	}
+	return r.reconcileVirtualNodeWithCloudMap(ctx, vNode)
+}
+
+func (r *cloudMapReconciler) reconcileVirtualNodeWithCloudMap(ctx context.Context, vNode *appmesh.VirtualNode) error {
+
+	if !r.isCloudMapEnabledForVirtualNode(vNode) {
+		return errors.Errorf("CloudMap not enabled for virtualNode: %s", vNode.Name)
```

Comment:
this shouldn't be error. when return error, it will backoff reconcile.
these are valid configurations, we should just return nil.